File: brikole/web/src/data/requests.ts
/** A client's own job requests. */

import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query'

import { api } from '@/data/client'
import type { Page, ProviderCity, Trade } from '@/data/types'

export type Urgency = 'today' | 'this_week' | 'flexible'
export type RequestStatus = 'open' | 'assigned' | 'done' | 'cancelled' | 'expired'

export interface RequestPhoto {
  id: number
  url: string
}

export interface ServiceRequest {
  id: number
  title: string
  description: string
  address: string
  latitude: number | null
  longitude: number | null
  trade: Trade
  city: ProviderCity
  urgency: Urgency
  status: RequestStatus
  budget_min_centimes: number | null
  budget_max_centimes: number | null
  /** How many tradesmen have answered. The loudest number on C2. */
  offers_count: number
  photos: RequestPhoto[]
  created_at: string
  expires_at: string | null
  cancelled_at: string | null
  cancel_reason: string | null
}

export interface NewRequestBody {
  trade_id: number
  city_id: number
  title: string
  description: string
  address: string
  latitude: number | null
  longitude: number | null
  urgency: Urgency
  budget_min_centimes: number | null
  budget_max_centimes: number | null
  photo_paths: string[]
}

export const REQUESTS_KEY = ['client', 'requests'] as const

export function useMyRequests() {
  return useQuery({
    queryKey: REQUESTS_KEY,
    queryFn: () => api<Page<ServiceRequest>>('/client/requests?per_page=50'),
    staleTime: 15_000,
  })
}

export function useMyRequest(requestId: number | null) {
  return useQuery({
    queryKey: [...REQUESTS_KEY, requestId],
    queryFn: () => api<ServiceRequest>(`/client/requests/${requestId}`),
    enabled: requestId !== null,
    staleTime: 15_000,
  })
}

export function useCreateRequest() {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: (body: NewRequestBody) =>
      api<ServiceRequest>('/client/requests', { method: 'POST', body }),
    onSuccess: () => queryClient.invalidateQueries({ queryKey: REQUESTS_KEY }),
  })
}

export function useCancelRequest() {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: ({ requestId, reason }: { requestId: number; reason?: string }) =>
      api<ServiceRequest>(`/client/requests/${requestId}/cancel`, {
        method: 'POST',
        body: { reason: reason ?? null },
      }),
    onSuccess: () => queryClient.invalidateQueries({ queryKey: REQUESTS_KEY }),
  })
}

export type OfferStatus = 'pending' | 'accepted' | 'rejected' | 'withdrawn' | 'expired'

export interface OfferProvider {
  id: number
  full_name: string
  avatar_url: string | null
  headline: string | null
  city: ProviderCity
  rating_avg: number
  rating_count: number
  jobs_done: number
  years_experience: number
}

export interface Offer {
  id: number
  price_centimes: number
  message: string
  available_from: string | null
  status: OfferStatus
  created_at: string
  provider: OfferProvider
}

export function useRequestOffers(requestId: number | null) {
  return useQuery({
    queryKey: [...REQUESTS_KEY, requestId, 'offers'],
    queryFn: () => api<Offer[]>(`/client/requests/${requestId}/offers`),
    enabled: requestId !== null,
    staleTime: 15_000,
  })
}

File: brikole/api/app/api/routes/requests.py
"""C1, C2 and C3 — a client's own job requests."""

from __future__ import annotations

from typing import Annotated

from fastapi import APIRouter, Depends, Query, status

from app.core.enums import Role
from app.deps import CurrentUser, DbSession, require_roles
from app.models.offer import Offer
from app.models.request import ServiceRequest
from app.schemas.catalog import TradeOut
from app.schemas.common import Page
from app.schemas.offer import OfferOut, OfferProviderOut
from app.schemas.provider import ProviderCityOut
from app.schemas.request import CancelRequestIn, NewRequestIn, RequestOut, RequestPhotoOut
from app.services.requests import RequestService

router = APIRouter(
    prefix="/client/requests",
    tags=["client"],
    dependencies=[Depends(require_roles(Role.CLIENT))],
)


@router.get("", response_model=Page[RequestOut])
def list_requests(
    user: CurrentUser,
    db: DbSession,
    page: Annotated[int, Query(ge=1)] = 1,
    per_page: Annotated[int, Query(ge=1, le=50)] = 20,
) -> Page[RequestOut]:
    """C2. Newest first — his own, and nobody else's."""
    rows, total = RequestService(db).list_own(user, page=page, per_page=per_page)
    return Page[RequestOut](
        items=[_out(row) for row in rows], total=total, page=page, per_page=per_page
    )


@router.post("", response_model=RequestOut, status_code=status.HTTP_201_CREATED)
def create_request(payload: NewRequestIn, user: CurrentUser, db: DbSession) -> RequestOut:
    """C1."""
    return _out(RequestService(db).create(user, payload))


@router.get("/{request_id}", response_model=RequestOut)
def get_request(request_id: int, user: CurrentUser, db: DbSession) -> RequestOut:
    """C3."""
    return _out(RequestService(db).get_own(user, request_id))


@router.get("/{request_id}/offers", response_model=list[OfferOut])
def list_offers(request_id: int, user: CurrentUser, db: DbSession) -> list[OfferOut]:
    """C3. Unsorted beyond newest first — the sorting a client wants (price,
    rating, soonest) is a decision he changes on the page, not a round trip."""
    return [_offer_out(offer) for offer in RequestService(db).list_offers(user, request_id)]


@router.post("/{request_id}/cancel", response_model=RequestOut)
def cancel_request(
    request_id: int, payload: CancelRequestIn, user: CurrentUser, db: DbSession
) -> RequestOut:
    return _out(RequestService(db).cancel(user, request_id, reason=payload.reason))


def _out(request: ServiceRequest) -> RequestOut:
    return RequestOut(
        id=request.id,
        title=request.title,
        description=request.description,
        address=request.address,
        latitude=request.latitude,
        longitude=request.longitude,
        trade=TradeOut.model_validate(request.trade),
        city=ProviderCityOut.model_validate(request.city),
        urgency=request.urgency,
        status=request.status,
        budget_min_centimes=request.budget_min_centimes,
        budget_max_centimes=request.budget_max_centimes,
        offers_count=request.offers_count,
        photos=[RequestPhotoOut.model_validate(photo) for photo in request.photos],
        created_at=request.created_at,
        expires_at=request.expires_at,
        cancelled_at=request.cancelled_at,
        cancel_reason=request.cancel_reason,
    )


def _offer_out(offer: Offer) -> OfferOut:
    provider = offer.provider
    return OfferOut(
        id=offer.id,
        price_centimes=offer.price_centimes,
        message=offer.message,
        available_from=offer.available_from,
        status=offer.status,
        created_at=offer.created_at,
        provider=OfferProviderOut(
            id=provider.id,
            full_name=provider.user.full_name,
            avatar_url=provider.user.avatar_url,
            headline=provider.headline,
            city=ProviderCityOut.model_validate(provider.city),
            rating_avg=provider.rating_avg,
            rating_count=provider.rating_count,
            jobs_done=provider.jobs_done,
            years_experience=provider.years_experience,
        ),
    )
